src/backend/api/learning_endpoints.py

```
@@ -432,20 +446,32 @@ def create_hint(request: HintCreateRequest, fb=Depends(_require_feedback_loop)):
+            )
+        except sqlite3.IntegrityError:
+            # BEGIN IMMEDIATE is a no-op on Postgres, so the duplicate check
+            # above no longer holds a write lock: a concurrent identical create
+            # can land between the SELECT and this INSERT. The UNIQUE constraint
+            # is the real guard — surface the loser as a retriable conflict,
+            # not a 500.
+            conn.rollback()
+            raise HTTPException(
+                status_code=409,
+                detail="An identical hint was just created — refresh to see it.",
+            )
```

Ruff (0.15.15)

[warning] Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)